### PR TITLE
[stable/traefik] Add KV Provider support in order to make Traefik HA

### LINKED ISSUE #6772: [stable/traefik] etcd provider
**Is this a request for help?**: No

---

**Is this a BUG REPORT or FEATURE REQUEST?** (choose one): Feature Request/Suggestion

**Version of Helm and Kubernetes**: Not relevant

**Which chart**: stable/traefik

**What happened**: 
I'm currently trying to deploy traefik in a cluster setup (which makes sense for a Kubernetes Ingress Controller I guess). I'm using ACME Certs and therefore I need to deploy either Consul or Etcd as a Backend.

There is already a PR open for using Consul as a backend. Me personally I would use Etcd as there is an operator by the CoreOS Team to spawn those cluster dynamically.
https://github.com/helm/charts/pull/5563

Somehow this PR seems to be stuck and adding Etcd as a backend to Helm will increase the complexity dramatically or at least break the other PR. So I don't know how to solve this for me except forking this repo, where I'm not a friend of forking without having the chance that it will be merged someday in the upstream again.

**What you expected to happen**:

I want to know how I could provide a PR using Etcd as a backend which has good chances to be merged.

**How to reproduce it** (as minimally and precisely as possible):

**Anything else we need to know**:

### PR DESCRIPTION
**Pull request by alwinmarkcf**
**What this PR does / why we need it**:
In order to be able to deploy Traefik as an **HA** Ingress Controller
This will add:
* the possibility to configure one of the supported KV Provider
* switch to store acme certs in that KV Provider or not
* A Kubernetes Job to create Keys and import certs from a given file into the KV, which can be Mounted threw any given claim, mounted like:
  ```
  acme:
    persistence:
      enabled: true
      existingClaim: acme-certs-pvc
  ```

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: Fixes: #6772

**Special notes for your reviewer**:
Copycated a lot from: https://github.com/helm/charts/pull/5563 so many thanks to: @siwyd

### CODE REVIEW

**Comment by benjigoldberg** [comment]
@dtomcej bump -- We'd love to see this get merged.

**Comment by benjigoldberg** [comment]
@alwinmarkcf one suggestion I do have is that we should give users the option to automatically install consul, etcd, whatever as a chart dependency and then take the service address of that k/v store. We could install a very small k/v store (1 tiny node) to handle traefiks configuration.

**Comment by alwinmarkcf** [comment]
@benjigoldberg I thought about that and decided to not add this as this make the whole setup much more complex. 

Me personally I'm creating "Master Charts" of everything with a custom values file an dependencies. This is why I created the storeconfig job as post-install hook. This allows one to install Etcd, Consul or whatever in the same masterchart and have both as dependencies. 
See:
https://docs.helm.sh/developing_charts/#chart-development-tips-and-tricks#complex-charts-with-many-dependencies

Where I prefer the deis solution as its more flexible. I could provide my full example with Etcd and Etcd Operator (where I would even not Install the Operator as dependecy but in a prerequisite step). This example than could be easily adapted for usage with: 

In my "Master Chart" I have e.g. this (currently monkeypatched chart) as dependency and I'm installing etcd threw the [Etcd Operator ](https://github.com/coreos/etcd-operator) like:
```
apiVersion: "etcd.database.coreos.com/v1beta2"
kind: "EtcdCluster"
metadata:
  name: {{ .Values.etcdCluster.name }}
  labels:
     app: {{ .Values.etcdCluster.name }}
  annotations:
     etcd.database.coreos.com/scope: clusterwide
spec:
  size: 3
  version: "3.1.8"
```

where this is my value file:

```
# Default values for traefik.
# This is a YAML-formatted file.
# Declare variables to be passed into your templates.
#
etcdCluster:
  name: traefik-etcd-cluster

traefik:
  replicas: 3

  cpuRequest: 200m
  memoryRequest: 500Mi
  cpuLimit: 900m
  memoryLimit: 1Gi

  ssl:
    enabled: true
    enforced: true

  rbac:
    enabled: true

  etcd:
    endpoint: "traefik-etcd-cluster-client:2379"

  acme:
    persistence:
      enabled: true
      existingClaim: acme-certs-pvc
    email: email@example
    staging: false
    logging: true
    challengeType: dns-01
    dnsProvider:
      name: route53
      route53:
        AWS_ACCESS_KEY_ID: adsfdasfadsf
    enabled: true

  service:
    nodePorts:
      http: 80
      https: 443

  proxyProtocol:
    enabled: true
    trustedIPs: []


  dashboard:
    enabled: true
    domain: traefik.example.com
    ingress:
      annotations:
        traefik.ingress.kubernetes.io/whitelist-source-range: ""

  deployment:
    hostPort:
      httpEnabled: true
      httpsEnabled: true

  gzip:
    enabled: true

  accessLogs:
    enabled: true
    format: json

  metrics:
    prometheus:
      enabled: true
```

**Comment by benjigoldberg** [comment]
@alwinmarkcf I'm ok with that approach. We can always decide to integrate default options later and users can do what you've suggested. It is pretty common in the charts repo to provide a default backing store or related service (See Kafka and Zookeeper for example, Jaeger and Elastic/Cassandra, etc). That being said, its probably better to start with things at the scope you have them and maybe work up to that.

File: stable/traefik/values.yaml
Comment (benjigoldberg) [inline stable/traefik/values.yaml]: should we comment out these blank keys and provide examples?

File: stable/traefik/values.yaml
Comment (benjigoldberg) [inline stable/traefik/values.yaml]: Providing a `# ref: <traefik-docs-on-kv-provider>` would be useful

File: stable/traefik/values.yaml
Comment (benjigoldberg) [inline stable/traefik/values.yaml]: You'll need to update the `README.md` with the new values documentation

File: stable/traefik/templates/configmap.yaml
Comment (benjigoldberg) [inline stable/traefik/templates/configmap.yaml]: are the TLS settings only valid for `etcd`?

File: stable/traefik/templates/storeconfig-job.yaml
Comment (benjigoldberg) [inline stable/traefik/templates/storeconfig-job.yaml:7]: Should this be a `pre-install` hook? We need this present before the Traefik pods come online, right? [The (admittedly sparse) documentation implies that it should be run before Traefik is started](https://docs.traefik.io/user-guide/cluster/#deploy-a-trfik-cluster). That being said, if you _were_ to deploy `etcd`, `consul`, etc alongside this I imagine a preinstall hook would get you into trouble. 

That

Comment (alwinmarkcf) [reply stable/traefik/templates/storeconfig-job.yaml:7]: Nope, if we make this pre-install we can't install in a master chart consul/etcd or sth. similar as the hook prevents further install process.

On the other hand traefik may fail once, but for me starting the server is slower than running the storeconfig job:

```
traefik-ingress-844f599647-bjcfg          1/1       Running     0          4d
traefik-ingress-844f599647-jxwxm          1/1       Running     0          4d
traefik-ingress-844f599647-kqhtm          1/1       Running     0          4d
```

Don't know if its just because of my cluster but this approach seems to be the most stable one.

The worst case which can happen is that the container gets rescheduled after the storeconfig job succeeded.

File: stable/traefik/values.yaml
Comment (alwinmarkcf) [reply stable/traefik/values.yaml]: Thanks I'll do

File: stable/traefik/templates/configmap.yaml
Comment (alwinmarkcf) [reply stable/traefik/templates/configmap.yaml]: No I just missed to generalize this one

**Comment by alwinmarkcf** [comment]
Except one thing, which is still to discuss I hopefully fixed every open review note. I did small commits, which I will squash as soon as I get a LGTM

File: stable/traefik/templates/storeconfig-job.yaml
Comment (benjigoldberg) [reply stable/traefik/templates/storeconfig-job.yaml:7]: makes sense to me 👍  -- thank you for the detailed response!

**Comment by alwinmarkcf** [comment]
@benjigoldberg thank you for your review :-)

File: stable/traefik/templates/configmap.yaml
Comment (dtomcej) [inline stable/traefik/templates/configmap.yaml]: This should be configurable. It is not defaulted to true when using the etcd provider.

**Review by dtomcej — CHANGES_REQUESTED** [review]

File: stable/traefik/templates/configmap.yaml
Comment (alwinmarkcf) [reply stable/traefik/templates/configmap.yaml]: Yeah absolutely. Very attentive! Fixed it

**Comment by alwinmarkcf** [comment]
So is everything good enough, so I can squash commits?

**Comment by alwinmarkcf** [comment]
So I squashed everything and updated the branch. If nothing is left todo, could somebody please merge this? 

@dtomcej  ???

**Comment by dtomcej** [comment]
Hello @alwinmarkcf 

Reviewed this PR with the team, and we have some feedback. We like the PR, but we need some changes to allow flexibility, and to prevent having features blocked by breaking changes.

1. The `storeconfig` and acme config changes are fine, as they do not rely on the kv provider or name.
2. We would like the template expanded from `[[{{ .Values.kvprovider.type }}]` to separate blocks for each provider. For now, only etcd and consul would have to be templated.

- This fits with the configuration style of traefik, and allows other features to be added later without having to break the chart.
- This would allow other users to extend the functionality by adding other kv providers
- This would allow users to use etcd for acme storage, and other provider for configuration in addition to kubernetes.

Thanks!

**Comment by alwinmarkcf** [comment]
I'm not sure if I understand you correctly.

Should it be possible having more than one Provider? What happens if I have
more than one, where are configs and acme certs are stored?

Could you provide some example configs you are planning to support in
future, so I get a better idea what you mean?

Thanks

Daniel Tomcej <notifications@github.com> schrieb am Fr., 3. Aug. 2018,
18:03:

**Comment by alwinmarkcf** [comment]
@dtomcej even though I was not sure if I understood you right, I changed this PR the way I thought you meant it. Hopefully its ok

File: stable/traefik/README.md
Comment (dtomcej) [inline stable/traefik/README.md]: The default should be `false` instead of `V2`

**Review by dtomcej — CHANGES_REQUESTED** [review]
Very good work! 👏 
Super close!
One final little documentation request

**Comment by dtomcej** [comment]
Weird that tests were running without an `ok-to-test`

**Comment by alwinmarkcf** [comment]
Yay :tada:

**Comment by siwyd** [comment]
@alwinmarkcf Cheers for this! I lacked any time whatsoever to continue what I started so many many thanks goes out to you.

**Comment by alwinmarkcf** [comment]
Unfortunately there seems to be an issue upgrading traffic. https://github.com/containous/traefik/issues/3754 

I'm checking this right now because I think I found the problem, so please keep this in mind until the issue is fixed or resolved.

**Comment by KlavsKlavsen** [comment]
How is kvprovider etcd ca, crt and key settings (in yaml) supposed to be set? I tried a ./etcd-client.crt (referring to the local host where I am running helm) - and that did not work. I also tried base64 encoding the files and putting them in yaml (starting with LS).. but that fails too.

**Comment by dtomcej** [comment]
@KlavsKlavsen It appears that the configuration references the files in the running pod.
If you want to provide the certs via the chart itself, you may want to open a PR to extend that functionality.

**Comment by KlavsKlavsen** [comment]
@dtomcej How do I get helm to mount files (I assume by somehow attaching a secret to it?) inside traefik and storeconfig-job pods? Is noone using traefik with TLS against a kvprovider ? sounds very odd.

**Pull request merged**
